[QA] Move e2e tests to vizro part two

l0uden opened the pull request:
## Description
In this PR I've moved the rest of the e2e dom elements tests fro vizro-qa to vizro repo.
Used `dash_br` fixture to do it and rewrote all of the tests and its helpers according to it.
Started to use CSS selectors for accessing the elements instead of XPATH. This approach made tests faster and stable.
Used original `selenium` only for the `statelessness tests`, because this tests are using two browsers and `dash_br` is not supporting it.


## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

    - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
    - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
    - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
    - I have not referenced individuals, products or companies in any commits, directly or indirectly.
    - I have not added data or restricted code in any commits, directly or indirectly.

petar-qb commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_parameters.py:
```diff
@@ -0,0 +1,40 @@
+    check_slider_value(dash_br, elem_id=cnst.RANGE_SLIDER_PARAMETERS, expected_start_value="4", expected_end_value="7")
+
+
+def test_none_parameter(dash_br):
+    page_select(
+        dash_br,
+        page_path=cnst.PARAMETERS_PAGE_PATH,
+        page_name=cnst.PARAMETERS_PAGE,
+        graph_id=cnst.BAR_GRAPH_ID,
+    )
+    # choose NONE parameter
+    select_dropdown_value(dash_br, value=1, dropdown_id=cnst.DROPDOWN_PARAMETERS_TWO)
+    check_graph_is_loading(dash_br, graph_id=cnst.BAR_GRAPH_ID)
```
I’ve noticed that many tests end with `check_graph_is_loading(dash_br, graph_id=GRAPH_ID)`, and while this is considered a valid test, it doesn't actually validate the core functionality. I understand that testing filtering or parameters can be challenging without relying on screenshot tests, and we should maybe reconsider how we approach these use cases. It might be worth exploring other ways to verify that the features are working as expected, without solely relying on graphical snapshots.

l0uden replied:
I think I've found a way here. Thanks for noticing!

petar-qb commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_statelessness.py:
I just have to add that understanding the implementation of the tests in this file was so easy and one of the main reasons is the code comments above the code blocks. 🚀

petar-qb commented on vizro-core/tests/tests_utils/e2e/asserts.py:
```diff
@@ -54,3 +56,18 @@ def assert_image_not_equal(image_one, image_two):
+def assert_files_equal(base_file, exported_file):
+    df_first = pd.read_csv(
+        base_file,
+        header=0,
+        on_bad_lines="skip",
```
Does anything fail if you remove these two properties? It shouldn't I guess 😄

l0uden replied:
Done

petar-qb commented on vizro-core/tests/tests_utils/e2e/asserts.py:
```diff
@@ -54,3 +56,18 @@ def assert_image_not_equal(image_one, image_two):
+        on_bad_lines="skip",
+    )
+    df_second = pd.read_csv(exported_file, header=0, on_bad_lines="skip")
+    result_diff = df_first[~df_first.apply(tuple, 1).isin(df_second.apply(tuple, 1))]
```
This won't catch if :

`base_file.csv` is:
```csv
Continent,Country
Asia,Japan
``` 

and `exported_file.csv` is:
```csv
Continent,Country
Asia,Japan
Europe,Germany
``` 

So, it doesn't look like it would fetch if the exported data is not filtered (but it should be).

This happens because the `result_diff` only contains the rows that exist in the `df_first` but not in the `df_second`. It doesn't contain the rows that exist in `df_second` but don't exist in the `df_first`. 

Something like this would work properly I guess:
```py
def assert_files_equal(base_file, exported_file):
    df_first = pd.read_csv(base_file, header=0, on_bad_lines="skip")
    df_second = pd.read_csv(exported_file, header=0, on_bad_lines="skip")

    # Check rows in base_file that are missing in exported_file
    missing_in_exported = df_first[~df_first.apply(tuple, 1).isin(df_second.apply(tuple, 1))]
    
    # Check rows in exported_file that are missing in base_file
    missing_in_base = df_second[~df_second.apply(tuple, 1).isin(df_first.apply(tuple, 1))]
    
    differences = pd.concat([missing_in_exported, missing_in_base])

    assert_that(
        differences.empty,
        equal_to(True),
        reason=f"{exported_file} is not equal to {base_file}\nDifference is:\n{differences}",  # TODO: Improve the message then
    )
```

l0uden replied:
Great catch!

petar-qb commented:
@l0uden congrats on the amazing job done!! 🥇 🚀 

Overall, this PR looks pretty good and most of the comments I left were minor suggestions 😃

l0uden commented on .github/workflows/test-e2e-vizro-dom-elements.yml:
```diff
@@ -18,26 +18,29 @@ env:
       fail-fast: false
       matrix:
-        config:
+        group: [1, 2, 3]
```
Yes

l0uden commented on vizro-core/tests/e2e/vizro/conftest.py:
```diff
@@ -58,11 +60,6 @@ def wait_for_callbacks(dash_br):
-        # "driver",
-        # "chromedriver_second_browser",
-    ]:
+    for driver_name in ["dash_br"]:
```
Done

l0uden commented on vizro-core/tests/e2e/vizro/custom_components/custom_components/dropdown.py:
```diff
@@ -53,5 +53,5 @@ def build(self):
-vm.Filter.add_type("selector", NewDropdown)
-vm.Parameter.add_type("selector", NewDropdown)
+vm.Filter.add_type("selector", CustomDropdown)
+vm.Parameter.add_type("selector", CustomDropdown)
```
No, deleted

l0uden commented on vizro-core/tests/e2e/vizro/dashboards/default/pages/dynamic_filters_pages.py:
Yes, I'm using all of them in the tests. It helps to access elements more easily.

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_custom_components.py:
```diff
@@ -0,0 +1,27 @@
+    )
+    dash_br.multiple_click(slider_value_path(elem_id=cnst.CUSTOM_RANGE_SLIDER_ID, value=4), 1)
+    check_graph_is_loading(dash_br, graph_id=cnst.SCATTER_CUSTOM_COMPONENTS_ID)
+    check_slider_value(dash_br, elem_id=cnst.CUSTOM_RANGE_SLIDER_ID, expected_start_value="4", expected_end_value="7")
```
Done

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_datepicker.py:
```diff
@@ -0,0 +1,64 @@
+    dash_br.wait_for_text_to_equal(
+        table_cell_value_path(table_id=cnst.TABLE_POP_RANGE_ID, row_number=2, column_number=1), "2016-05-17T00:00:00"
+    )
+    check_table_rows_number(dash_br, table_id=cnst.TABLE_POP_RANGE_ID, expected_rows_num=4)
```
Done

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_datepicker.py:
```diff
@@ -0,0 +1,64 @@
+    dash_br.wait_for_element('div[data-calendar="true"]')
+    dash_br.multiple_click('button[aria-label="2 April 2018"]', 1)
+    # check that one of the bars change color from blue to orange
+    dash_br.wait_for_element(f"div[id='{cnst.BAR_CUSTOM_ID}'] g:nth-of-type(14) path[style*='(255, 165, 0)'")
```
Done

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_dynamic_data.py:
```diff
@@ -1,12 +1,48 @@
 from e2e.vizro.waiters import callbacks_finish_waiter
 
 
+def dynamic_filters_data_config_manipulation(key, set_value=None):
+    with open(cnst.DYNAMIC_FILTERS_DATA_CONFIG) as file:
+        data = yaml.safe_load(file)
+        data[key] = set_value
+    with open(cnst.DYNAMIC_FILTERS_DATA_CONFIG, "w") as file:
+        yaml.dump(data, file, default_flow_style=False)
```
Yes!

l0uden replied:
Actually made it a bit different with 'r+', because 'w+' erased all the data

petar-qb replied:
Since the content length (number of lines in the file) is the same before and after this function, both `r+` and `w+` will work for you. Here's a useful diagram that I've referred to when I forget the difference between these options.

<img width="672" alt="image" src="https://github.com/user-attachments/assets/fd6d5226-473f-47bf-946f-0363e424ad97" />

l0uden replied:
Thanks for this very useful diagram!

I can't truncate the file during reading, because I need to change the value there, and if it will be truncated I will get 'NoneType' error. That is why I've used 'r+'

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_dynamic_data.py:
```diff
@@ -54,3 +90,244 @@ def test_data_dynamic_parametrization(dash_br, cache, slider_id):
+
+
+@rewrite_dynamic_filters_data_config
+def test_sliders_filters(dash_br):
```
Done

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_interactions.py:
```diff
@@ -0,0 +1,28 @@
+    dash_br.click_at_coord_fractions(f"#{cnst.SCATTER_INTERACTIONS_ID} path:nth-of-type(20)", 0, 1)
+    check_graph_is_loading(dash_br, cnst.BOX_INTERACTIONS_ID)
+
+    # select 'setosa' in fropdown filter
```
Done

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_filter_and_param_config.py:
```diff
@@ -0,0 +1,18 @@
+from e2e.vizro.paths import categorical_components_value_path
+
+
+def test_filter_and_parameter(dash_br):
+    page_select(
+        dash_br,
+        page_path=cnst.FILTER_AND_PARAM_PAGE_PATH,
+        page_name=cnst.FILTER_AND_PARAM_PAGE,
+        graph_id=cnst.BOX_FILTER_AND_PARAM_ID,
+    )
+    dash_br.multiple_click(categorical_components_value_path(elem_id=cnst.RADIO_ITEMS_FILTER_AND_PARAM, value=1), 1)
+    check_graph_is_loading(dash_br, graph_id=cnst.BOX_FILTER_AND_PARAM_ID)
+    dash_br.wait_for_text_to_equal(".gtitle", "red")
+    select_dropdown_value(dash_br, value=1, dropdown_id=cnst.DROPDOWN_FILTER_AND_PARAM)
+    check_graph_is_loading(dash_br, graph_id=cnst.BOX_FILTER_AND_PARAM_ID)
```
Page with filter and parameter in the same config

l0uden commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_parameters.py:
```diff
@@ -0,0 +1,40 @@
+from e2e.vizro.paths import slider_value_path
+
+
+def test_sliders_state(dash_br):
+    page_select(
+        dash_br,
+        page_path=cnst.PARAMETERS_PAGE_PATH,
+        page_name=cnst.PARAMETERS_PAGE,
+        graph_id=cnst.BAR_GRAPH_ID,
+    )
+    dash_br.multiple_click(slider_value_path(elem_id=cnst.SLIDER_PARAMETERS, value=3), 1)
+    check_graph_is_loading(dash_br, graph_id=cnst.BAR_GRAPH_ID)
+    dash_br.multiple_click(slider_value_path(elem_id=cnst.RANGE_SLIDER_PARAMETERS, value=4), 1)
+    check_graph_is_loading(dash_br, graph_id=cnst.HISTOGRAM_GRAPH_ID)
+
+    page_select(dash_br, page_path=cnst.FILTERS_PAGE_PATH, page_name=cnst.FILTERS_PAGE, graph_id=cnst.SCATTER_GRAPH_ID)
+    page_select(
+        dash_br,
+        page_path=cnst.PARAMETERS_PAGE_PATH,
+        page_name=cnst.PARAMETERS_PAGE,
+        graph_id=cnst.BAR_GRAPH_ID,
+    )
+
+    check_slider_value(dash_br, expected_end_value="0.4", elem_id=cnst.SLIDER_PARAMETERS)
+    check_slider_value(dash_br, elem_id=cnst.RANGE_SLIDER_PARAMETERS, expected_start_value="4", expected_end_value="7")
+
```
Added

l0uden commented:
> @l0uden congrats on the amazing job done!! 🥇 🚀
> 
> Overall, this PR looks pretty good and most of the comments I left were minor suggestions 😃

Thanks, @petar-qb ! 

You can re-review now.

petar-qb commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_dynamic_data.py:
```diff
@@ -54,3 +90,244 @@ def test_data_dynamic_parametrization(dash_br, cache, slider_id):
+    )
+
+    # Check that "versicolor" and "virginica" is the only listed options
+    check_selected_checklist(
+        dash_br,
+        checklist_id=cnst.CHECKLIST_DYNAMIC_FILTER_ID,
+        select_all_status=False,
+        options_value_status=[
+            {"value": 1, "status": False, "value_name": "ALL"},
+            {"value": 2, "status": True, "value_name": "versicolor"},
+            {"value": 3, "status": False, "value_name": "virginica"},
+        ],
+    )
```
Great, let's link this conversation somewhere in the "SelectALL" PR so that we don't forget 😄

l0uden replied:
Added to the description

petar-qb commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_filter_and_param_config.py:
```diff
@@ -0,0 +1,19 @@
+from e2e.vizro.paths import categorical_components_value_path
+
+
+def test_filter_and_parameter(dash_br):
+    """Testing filter and parameter on the same page."""
+    page_select(
+        dash_br,
+        page_path=cnst.FILTER_AND_PARAM_PAGE_PATH,
+        page_name=cnst.FILTER_AND_PARAM_PAGE,
+        graph_id=cnst.BOX_FILTER_AND_PARAM_ID,
+    )
+    dash_br.multiple_click(categorical_components_value_path(elem_id=cnst.RADIO_ITEMS_FILTER_AND_PARAM, value=1), 1)
+    check_graph_is_loading(dash_br, graph_id=cnst.BOX_FILTER_AND_PARAM_ID)
+    dash_br.wait_for_text_to_equal(".gtitle", "red")
+    select_dropdown_value(dash_br, value=1, dropdown_id=cnst.DROPDOWN_FILTER_AND_PARAM)
+    check_graph_is_loading(dash_br, graph_id=cnst.BOX_FILTER_AND_PARAM_ID)
```
The reason why I haven't understood this in the first place was because the RadioItems Parameter value is changed with `dash_br.multiple_click(...)` and the Dropdown Filter value with the `select_dropdown_value(...)`. 😄 

Let's add a few comments here and maybe a check that the title was blue at the beginning of this test.

l0uden replied:
Done

petar-qb reviewed (approved):
Fantastically addressed comments! I really like the code changes and especially the comments you added! 🚀 Well done

l0uden commented on vizro-core/tests/e2e/vizro/conftest.py:
```diff
@@ -49,20 +38,26 @@ def dash_br_driver(dash_br, request):
-        "dash_br",
-        #  TODO: uncomment for the full scope of tests
-        # "driver",
-        # "chromedriver_second_browser",
-    ]:
-        if (driver := request.node.funcargs.get(driver_name)) is not None:
-            make_teardown(driver)
+    make_teardown(dash_br)
```
Yes, sure. Great that you've saw it!

l0uden commented:
> Fantastically addressed comments! I really like the code changes and especially the comments you added! 🚀 Well done

Thanks Petar!
Now tests looks even better!

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_dynamic_data.py:
```diff
@@ -54,3 +99,447 @@ def test_data_dynamic_parametrization(dash_br, cache, slider_id):
+        page_name=cnst.DYNAMIC_FILTERS_CATEGORICAL_PAGE,
+        graph_id=cnst.BOX_DYNAMIC_FILTERS_ID,
+    )
+
```
The comments are so good, I would have more.

l0uden replied:
I will resolve all comments about comments because I've added it in all tests.

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_dynamic_data.py:
```diff
@@ -54,3 +99,447 @@ def test_data_dynamic_parametrization(dash_br, cache, slider_id):
+
+@rewrite_dynamic_filters_data_config
+def test_dropdown_filter_multi(dash_br):
+    """Initial selected value is 'ALL'."""
```
Agree with @petar-qb , super good to read. Love the docstring with the starting state, maybe even write the end state we want to achieve as well?

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_dynamic_data.py:
```diff
@@ -54,3 +99,447 @@ def test_data_dynamic_parametrization(dash_br, cache, slider_id):
+        page_name=cnst.DYNAMIC_FILTERS_CATEGORICAL_PAGE,
+        graph_id=cnst.BOX_DYNAMIC_FILTERS_ID,
+    )
+
```
same here

maxschulz-COL replied:
and so on, ie just a comment above all actions, and describe also end-state (+ maybe what the test is testing, beyond just the test name)

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_interactions.py:
```diff
@@ -0,0 +1,28 @@
+
+    # click on the 'setosa' data in scatter graph
+    dash_br.click_at_coord_fractions(f"#{cnst.SCATTER_INTERACTIONS_ID} path:nth-of-type(20)", 0, 1)
+    check_graph_is_loading(dash_br, cnst.BOX_INTERACTIONS_ID)
```
Is it necessary to check the correct filtering here? Or what is this checking against?

l0uden replied:
Added this check

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_parameters.py:
```diff
@@ -0,0 +1,57 @@
+from e2e.vizro.paths import slider_value_path
+
+
+def test_sliders_state(dash_br):
```
add some commentary here as well, where do we start, where do we end?

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_statelessness.py:
```diff
@@ -0,0 +1,114 @@
+from selenium.webdriver.support.wait import WebDriverWait
+
+
+def test_parameters_title(chrome_driver, dash_br):
```
wow is this testing it for a single server, but multiple people using it?

l0uden replied:
Yes

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_statelessness.py:
```diff
@@ -0,0 +1,114 @@
+from selenium.webdriver.support.wait import WebDriverWait
+
+
+def test_parameters_title(chrome_driver, dash_br):
```
small explanation of what is initial and final state?

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_table.py:
```diff
@@ -0,0 +1,60 @@
+from e2e.vizro.paths import categorical_components_value_path, slider_value_path, table_cell_value_path
+
+
+def test_filters(dash_br):
```
same here

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_table.py:
```diff
@@ -0,0 +1,60 @@
+    dash_br.wait_for_text_to_equal(table_cell_value_path(table_id=cnst.TABLE_ID, row_number=3, column_number=8), "108")
+
+
+def test_interactions(dash_br):
```
here too

maxschulz-COL commented on vizro-core/tests/e2e/vizro/test_dom_elements/test_table_ag_grid.py:
```diff
@@ -0,0 +1,61 @@
+from e2e.vizro.paths import categorical_components_value_path, slider_value_path, table_ag_grid_cell_value_path
+
+
+def test_filters(dash_br):
```
will stop commenting now

maxschulz-COL commented on vizro-core/tests/tests_utils/e2e/asserts.py:
```diff
@@ -54,3 +56,22 @@ def assert_image_not_equal(image_one, image_two):
         pass
+
+
+def assert_files_equal(base_file, exported_file):
```
there must be an easier way to compare two DFs no?

l0uden replied:
There is, but there's no easier way if we want to see what is actually different.

maxschulz-COL reviewed (approved):
Skimmed over it, thanks @petar-qb for the detailed review.

Feel free to ignore my comments, I would be happy to merge as is. It looks really good, and is quite readable!
